src/modules/accounting/accounting.py

```diff
@@ -0,0 +1,239 @@
+        lido_validators = self.w3.lido_validators.get_lido_validators(blockstamp)
+
+        count = len(lido_validators)
+        total_balance = Gwei(sum(int(validator.balance) for validator in lido_validators))
+
+        logger.info({'msg': 'Calculate consensus lido state.', 'value': (count, total_balance)})
+        return count, total_balance
+
+    def _get_last_withdrawal_request_to_finalize(self, blockstamp: ReferenceBlockStamp) -> int:
+        chain_config = self.get_chain_config(blockstamp)
+        frame_config = self.get_frame_config(blockstamp)
+
+        is_bunker = self._is_bunker(blockstamp)
+        withdrawal_vault_balance = self.w3.lido_contracts.get_withdrawal_balance(blockstamp)
+        el_rewards_vault_balance = self.w3.lido_contracts.get_el_vault_balance(blockstamp)
+        finalization_share_rate = self._get_finalization_shares_rate(blockstamp)
+
+        withdrawal_service = Withdrawal(self.w3, blockstamp, chain_config, frame_config)
+
+        last_wr_id = withdrawal_service.get_next_last_finalizable_id(
+            is_bunker,
+            finalization_share_rate,
+            withdrawal_vault_balance,
+            el_rewards_vault_balance,
+        )
+        logger.info({'msg': 'Calculate last withdrawal id to finalize.', 'value': last_wr_id})
+        return last_wr_id
+
+    @lru_cache(maxsize=1)
+    def _get_finalization_shares_rate(self, blockstamp: ReferenceBlockStamp) -> int:
+        simulation = self.get_rebase_after_report(blockstamp)
+        shares_rate = simulation.post_total_pooled_ether * SHARE_RATE_PRECISION_E27 // simulation.post_total_shares
+        logger.info({'msg': 'Calculate shares rate.', 'value': shares_rate})
+        return shares_rate
+
+    def get_rebase_after_report(self, blockstamp: ReferenceBlockStamp, cl_only=False) -> LidoReportRebase:
+        chain_conf = self.get_chain_config(blockstamp)
+        frame_config = self.get_frame_config(blockstamp)
+
+        last_ref_slot = self.report_contract.functions.getLastProcessingRefSlot().call(
+            block_identifier=blockstamp.block_hash,
+        )
+
+        if last_ref_slot:
+            slots_elapsed = blockstamp.ref_slot - last_ref_slot
+        else:
+            slots_elapsed = blockstamp.ref_slot - frame_config.initial_epoch * chain_conf.slots_per_epoch
+
+        validators_count, cl_balance = self._get_consensus_lido_state(blockstamp)
+
+        timestamp = chain_conf.genesis_time + blockstamp.ref_slot * chain_conf.seconds_per_slot
+
+        handle_report_data = (
+            timestamp,  # _reportTimestamp
+            slots_elapsed * chain_conf.seconds_per_slot,  # _timeElapsed
+            validators_count,  # _clValidators
+            Web3.to_wei(cl_balance, 'gwei'),  # _clBalance
+            self.w3.lido_contracts.get_withdrawal_balance(blockstamp),  # _withdrawalVaultBalance
+            0 if cl_only else self.w3.lido_contracts.get_el_vault_balance(blockstamp),  # _elRewardsVaultBalance
```

Comment:
let's move out `cl_only` flag (maybe split into two different methods)
for this context it's a kind of hack

Reply:
Not sure if it will increase readability